test/shmemx/shmemx_malloc_with_hints.c

```diff
@@ -0,0 +1,111 @@
+/*
+ *  Copyright (c) 2017 Intel Corporation. All rights reserved.
+ *  This software is available to you under the BSD license below:
+ *
+ *      Redistribution and use in source and binary forms, with or
+ *      without modification, are permitted provided that the following
+ *      conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <stdint.h>
+#include <shmem.h>
+#include <shmemx.h>
+
+#define N 128
+#define MAX(a,b) ((a) > (b)) ? a : b
+#define WRK_SIZE MAX(N/2+1, SHMEM_REDUCE_MIN_WRKDATA_SIZE)
+
+#define NO_TESTS 4
+
+long pSync[SHMEM_ALLTOALL_SYNC_SIZE];
+int pWrk[WRK_SIZE];
+
+
+int *dst, *src;
+
+int sumtoall_with_malloc_hint(long hint, int mype, int npes)
+{
+    int failed = 0;
+    int i;
+
+    if (mype == 0)
+        printf("Sum reduction operation started\n");
+
+    dst = (int *)shmemx_malloc_with_hints(N * sizeof(*dst), hint);
+    src = (int *)shmemx_malloc_with_hints(N * sizeof(*src), hint);
+
+    for (i = 0; i < npes; i++) {
+        src[i] = mype;
+        dst[i] = -1;
+    }
+
+    shmem_barrier_all();
+    shmem_int_sum_to_all(dst, src, N, 0, 0, npes, pWrk, pSync);
+
+    if(mype == 0){
+        for (i = 0; i < npes; i++) {
+            int expected;
+            expected = npes * (npes-1)/2;
+            if (dst[i] != expected) {
+                printf("[%d] dst[%d] = %d, expected %d\n", mype, i, dst[i], expected);
+                failed = 1;
+            }
+        }
+    }
+
+    if (failed)
+        return 0;
+    return 1;
+}
+
+
+int main(int argc, char **argv) {
+    int npes, i, mype;
+    long malloc_hint;
+    int passed = 0;
+    int fail = 0;
+
+    shmem_init();
+
+    npes = shmem_n_pes();
+    mype = shmem_my_pe();
+
+    for (i = 0; i < SHMEM_ALLTOALL_SYNC_SIZE; i++)
+        pSync[i] = SHMEM_SYNC_VALUE;
+
+    passed = sumtoall_with_malloc_hint(malloc_hint, mype, npes);
```

Comment:
Looks like `malloc_hint` is not initialized here?

Reply:
aah yes good catch! I will remove that

Reply:
Looks like you meant to instantiate different hints combos before each call, right?

Reply:
I might suggest 5 hints:
1. `0`
2. `SHMEMX_MALLOC_ATOMICS_REMOTE`
3. `SHMEMX_MALLOC_SIGNAL_REMOTE`
4. `SHMEMX_MALLOC_ATOMICS_REMOTE | SHMEMX_MALLOC_SIGNAL_REMOTE`
5. Something invalid.